tools/fm-editor/FeatureModelEditor.cpp

```diff
@@ -0,0 +1,290 @@
+    : QMainWindow(Parent), Ui(new Ui::FeatureModelEditor) {
+
+  Ui->setupUi(this);
+  Ui->textEdit->setReadOnly(true);
+  auto *Highliter =
+      new QSourceHighlite::QSourceHighliter(Ui->textEdit->document());
+  Highliter->setCurrentLanguage(QSourceHighlite::QSourceHighliter::CodeCpp);
+  QObject::connect(Ui->loadModel, &QPushButton::pressed, this,
+                   &FeatureModelEditor::loadGraph);
+  QObject::connect(Ui->actionAddFeature, &QAction::triggered, this,
+                   &FeatureModelEditor::featureAddDialog);
+  connect(Ui->actionSave, &QAction::triggered, this, &FeatureModelEditor::save);
+  connect(Ui->addSource, &QPushButton::pressed, this,
+          &FeatureModelEditor::addSource);
+  connect(Ui->addSourceFile, &QPushButton::pressed, this,
+          &FeatureModelEditor::addSourceFile);
+}
+
+/// Display the information of a Feature
+void FeatureModelEditor::loadFeature(const vara::feature::Feature *Feature) {
+  auto FeatureString = Feature->toString();
+  Ui->featureInfo->setText(QString::fromStdString(FeatureString));
+}
+
+/// Get a Feature from an Index of the TreeView and display its information.
+void FeatureModelEditor::loadFeatureFromIndex(const QModelIndex &Index) {
+  if (Index.isValid()) {
+    auto Item = static_cast<FeatureTreeItem *>(Index.internalPointer())
+                    ->child(Index.row());
+    if (Item->getKind() ==
+        vara::feature::FeatureTreeNode::NodeKind::NK_FEATURE) {
+      loadFeature(static_cast<FeatureTreeItemFeature *>(Item)->getItem());
+    }
+  }
+}
+/// Clear all Fields that should be emptied when loading a new Model
+void FeatureModelEditor::clean() {
+  SavePath.clear();
+  Repository.clear();
+  Ui->tabWidget->clear();
+  Ui->sources->clear();
+  Ui->textEdit->clear();
+  Ui->sourcesLable->setText("Sources for: ");
+  Ui->featureInfo->clear();
+}
+
+/// Load the Feature Model at the Path in ModelFile field and build the Views
+void FeatureModelEditor::loadGraph() {
+  clean();
+  ModelPath = Ui->ModelFile->text();
+  FeatureModel = vara::feature::loadFeatureModel(ModelPath.toStdString());
+  if (!FeatureModel) {
+    QString const Path = QFileDialog::getOpenFileName(
+        this, tr("Open Model"), "/home", tr("XML files (*.xml)"));
+    if (Path.isEmpty()) {
+      return;
+    }
+    Ui->ModelFile->setText(Path);
+    FeatureModel = vara::feature::loadFeatureModel(Path.toStdString());
+  }
+  // create Graph view
+  buildGraph();
+  // create Tree View
+  buildTree();
+  Ui->tabWidget->addTab(Graph.get(),"GraphView");
+  Ui->tabWidget->addTab(TreeView,"TreeView");
+  connect(Ui->sources,&QComboBox::currentTextChanged, this,&FeatureModelEditor::loadSource);
+  Ui->actionSave->setEnabled(true);
+  Ui->actionAddFeature->setEnabled(true);
+}
+
+/// Build the Treeview
+void FeatureModelEditor::buildTree() {
+  TreeView = new QTreeView();
+  TreeModel =
+      std::make_unique<FeatureTreeViewModel>(FeatureModel.get(), TreeView);
```

Comment:
If we own this interface, we should use a unique_ptr for TreeView instead of passing objects allocated with new directly.